File: python/parmana/api/execution_api.py
"""
Parmana Execution API.

Execute Business Transactions through the Parmana Runtime.
"""

from __future__ import annotations

from typing import Any

from parmana.config.transport import Transport
from parmana.models.business_transaction import BusinessTransaction
from parmana.models.trust_record import ExecutionTrustRecord
from parmana.serialization import encode


class ExecutionApi:
    """
    Execution API.

    Responsibilities
    ----------------
    - Runtime health check
    - Execute Business Transactions

    This API does NOT:
    - verify trust records
    - replay executions
    - generate receipts
    - validate policies
    """

    def __init__(
        self,
        transport: Transport,
    ) -> None:
        self._transport = transport

    def health(
        self,
    ) -> dict[str, Any]:
        """
        Returns the Runtime health status.
        """

        return self._transport.send(
            method="GET",
            path="/health",
        )

    def execute(
        self,
        transaction: BusinessTransaction,
    ) -> ExecutionTrustRecord:
        """
        Execute a Business Transaction.

        Parameters
        ----------
        transaction:
            Business Transaction submitted to the Parmana Runtime.

        Returns
        -------
        Execution Trust Record.
        """

        return self._transport.send(
            method="POST",
            path="/execute",
            body=encode(transaction),
            response_model=ExecutionTrustRecord,
        )

File: typescript/src/client/ExecutionApi.ts
/**
 * Parmana Execution API.
 *
 * Execute Business Transactions through the Parmana Runtime.
 */

import type {
  BusinessTransaction,
  ExecutionTrustRecord,
} from "../models/index.js";

import type {
  Transport,
} from "../config/Transport.js";

/**
 * Execution API.
 *
 * Responsibilities
 * ----------------
 * - Runtime health check
 * - Execute Business Transactions
 *
 * This API does NOT:
 * - verify trust records
 * - replay executions
 * - generate receipts
 * - validate policies
 */
export class ExecutionApi {
  constructor(
    private readonly transport: Transport,
  ) {}

  /**
   * Returns the Runtime health status.
   */
  public async health(): Promise<unknown> {
    const response =
      await this.transport.send({
        method: "GET",
        path: "/health",
      });

    return response.body;
  }

  /**
   * Execute a Business Transaction.
   */
  public async execute(
    transaction: BusinessTransaction,
  ): Promise<ExecutionTrustRecord> {
    const response =
      await this.transport.send<ExecutionTrustRecord>({
        method: "POST",
        path: "/execute",
        body: transaction,
      });

    return response.body;
  }
}